docs(CSS): update scroll-timeline shorthand property

dipikabh opened the pull request:
<!-- 🙌 Thanks for contributing to MDN Web Docs. Adding details below will help us to merge your PR faster. -->

### Summary

- The shorthand and longhand properties, [scroll-timeline](https://developer.mozilla.org/en-US/docs/Web/CSS/scroll-timeline), [scroll-timeline-name](https://developer.mozilla.org/en-US/docs/Web/CSS/scroll-timeline-name),  
[scroll-timeline-axis](https://developer.mozilla.org/en-US/docs/Web/CSS/scroll-timeline-axis), are still behind the `layout.css.scroll-linked-animations.enabled` preference in Fx111. [UPDATE: this is actually the `layout.css.scroll-driven-animations.enabled` preference.]
- The order of name and axis is now fixed in the shorthand property via the linked bugzilla bug. This PR updates that part of content.
- The PR also adds a section title for the example in the "Examples" section.


### Supporting details

Doc issue tracker: https://github.com/mdn/content/issues/24396

Bug tracker: https://bugzilla.mozilla.org/show_bug.cgi?id=1804573

Spec: https://w3c.github.io/csswg-drafts/scroll-animations/#scroll-timeline-shorthand

dipikabh committed:
update scroll-timeline behavior

dipikabh committed:
Merge branch 'main' into fx111-scroll-timeline

dipikabh committed:
merge css code in one block

dipikabh committed:
Merge branch 'main' into fx111-scroll-timeline

estelle commented on files/en-us/web/css/scroll-timeline/index.md:
```diff
@@ -133,6 +133,10 @@ The CSS below defines a square that rotates in alternate directions according to
 }
 
+#stretcher {
+  height: 600px;
```
```suggestion
  height: 600px;
  background: #dedede;
```
just to make it visible

estelle commented on files/en-us/web/css/scroll-timeline/index.md:
```diff
@@ -49,7 +47,7 @@ scroll-timeline: horizontal;
 ### Values
 
-A `<scroll-timeline-name>` value followed by a `<scroll-timeline-axis>` value may be applied to the container element in any order.
+A `<scroll-timeline-name>` value followed by a `<scroll-timeline-axis>` value may be applied to the container element in that fixed order.
```
we should add a note that if only one value is included, and the name of the timeline is a possible value of scroll-timeline-axis, it will fail.

example: https://codepen.io/estelle/pen/abaVZeP

dipikabh replied:
So I checked the expected behavior with the Fx engineer. It seems the only case that will fail is when the order of values is 'axis name'. If only one value is specified, even if that's the axis keyword, the shorthand property will regard it as a name. The shorthand property only expects a name at the least. So I've worded the note accordingly.

estelle reviewed (commented):
I suggested adding a warning or note about scroll-timeline-name not being a potential value of scroll-timeline-axis. Also, it helps to add a bg color to stretcher and border or outline on container (optional, but i think users will better understand what is going one)

estelle commented:
I also recommend making the animation timing duration 1ms so users can see if it's working or not. with 5s, it can be coming from animation name or timeline. The difference is stark at 0.001s as you can't see the animation if timeline isn't supported.

estelle commented:
https://github.com/mdn/browser-compat-data/issues/19086

dipikabh committed:
Merge branch 'main' into fx111-scroll-timeline

dipikabh committed:
Merge branch 'main' into fx111-scroll-timeline

dipikabh commented:
Thanks a lot, @estelle for investigating the expected behavior since last night and troubleshooting the flag name 🙌

dipikabh committed:
fix desc, note, example

dipikabh committed:
fix desc, note, example in name page

dipikabh committed:
fix desc, note, example in axis page

dipikabh committed:
update experimental relnote

dipikabh committed:
remove blank line

dipikabh commented:
@estelle, thanks to your code help, I've now also added to this PR updates to the related longhand property pages.

- On the shorthand property page, I have removed descriptions of the values of the longhand properties. I've replaced these with links to the respective longhand property pages to avoid duplication of content.
- The new syntax for `scroll-timeline` shorthand as per the [spec](https://w3c.github.io/csswg-drafts/scroll-animations/#scroll-timeline-shorthand) is now:
`[ <'scroll-timeline-name'> <'scroll-timeline-axis'>? ]#`
This is yet to reflect in the "Formal syntax" section. I'll check mdn/data.
- The name of the spec changed from "scroll-linked animations" to "scroll-driven animations". And so the preference flag for this feature in Fx changed from `layout.css.scroll-linked-animations.enabled` to `layout.css.scroll-driven-animations.enabled`. Thanks for discovering this! I have updated the experimental features file to indicate this flag name change.

dipikabh commented:
Hello @captainbrosset, I have a BCD-related question for this CSS `scroll-timeline` property in Chrome.

Per https://chromestatus.com/feature/6752840701706240, I see the implementation status is "Proposed". However, the feature is working on 111 and 112 beta with "Experimental Web Platform Features" enabled.

https://github.com/mdn/browser-compat-data/issues/19086 mentions that the property was added to Chrome in version 40 (experimental).

Would it be possible for you to check the version it was first added (even if in experimental mode) in Chrome and Edge? And if a different flag was needed in the older releases?

estelle commented:
> [mdn/browser-compat-data#19086](https://github.com/mdn/browser-compat-data/issues/19086) mentions that the property was added to Chrome in version 40 (experimental).
> 

My bad. I did not update the drop down. The default value of a drop down is "added in chrome 40". 

the issue is that scroll-timeline does work, with flags, in chrome, and the BCD doesn't reflect this. scroll-timeline may not have even been a suggested property way back at 40 release time ;)

estelle commented on files/en-us/web/css/scroll-timeline/index.md:
```diff
@@ -116,47 +94,46 @@ The height of the container is set to 300px and we also set the container to cre
+{{EmbedLiveSample("Creating_a_scroll_driven_animation", "100%", "320px")}}
 
-{{EmbedLiveSample("Setting a scroll timeline", "100%", "320px")}}
+The square animates as you scroll, and the animation duration when using `scroll-timeline` really depends on the scroll speed (nevertheless, the `animation-duration` property has been defined so you can make out the scroll-driven animation).
```
```suggestion
The square animates as you scroll, and the animation duration when using `scroll-timeline` really depends on the scroll speed. If `animation` is supported, but `scroll-timeline` is not supported, the animation still occurs, lasting the duration as defined in the required [`animation-duration`](/en-US/docs/Web/CSS/animation-duration) property or `animation-duration` component of the [`animation`](/en-US/docs/Web/CSS/animation) shorthand property.
```

I need to check this. In my testing, 0s didn't work, but I haven't read in the spec that it's not supposed to.

estelle replied:
i am seeing that duration is treated as auto, but am not finding that it is "required", but, when I set time to 0ms in https://codepen.io/estelle/pen/zYJPYQm, it is definitely NOT working in any browser

estelle reviewed (commented):
Only 2 required changes. all the others are optional/suggestions/questions. I need to look up animation duration before going live with the last suggestion.

dipikabh commented:
Thanks for the feedback, Estelle. Incorporating the fixes now.

dipikabh committed:
Merge branch 'main' into fx111-scroll-timeline

dipikabh committed:
fix review comments

estelle commented on files/en-us/web/css/scroll-timeline-axis/index.md:
```diff
@@ -9,11 +9,12 @@ browser-compat: css.properties.scroll-timeline-axis
-The [`scroll-timeline-name`](/en-US/docs/Web/CSS/scroll-timeline-name) property defines a name that can be used to identify an element as the source of a scroll-timeline; the `scroll-timeline-axis` is used to indicate whether the horizontal or vertical scrollbar in that element provides the timeline.
-Note that if the element does not display a scrollbar in the axis dimension, then no timeline will be created.
+In a scroll-timeline animation, the animation progress is determined by a scrollbar position rather than time or some other dimension.
+The [`scroll-timeline-name`](/en-US/docs/Web/CSS/scroll-timeline-name) property defines a name that can be used to identify an element as the source of a scroll-timeline; the `scroll-timeline-axis` defines the horizontal or vertical scrollbar in the container that will provide the timeline.
```
```suggestion
The [`scroll-timeline-name`](/en-US/docs/Web/CSS/scroll-timeline-name) property defines a name that can be used to identify an element as the source of a scroll-timeline; the `scroll-timeline-axis` defines the direction of scrollbar in the container that will provide the timeline.
```

estelle replied:
not longer limited to these two values.

estelle commented on files/en-us/web/css/scroll-timeline-axis/index.md:
```diff
@@ -53,11 +54,13 @@ Allowed values for the scrollbar axis are:
+In this example, a scroll timeline named `squareTimeline` is defined using the `scroll-timeline-name` property on the element with the ID `container`, and then this name is applied to the animation on the `#square` element using `animation-timeline: squareTimeline`.
 
-To demonstrate the effect of `scroll-timeline-axis`, the horizontal (non-default) scrollbar is used in this example to control the rotation of the square.
+To demonstrate the effect of `scroll-timeline-axis`, the inline (non-default) scrollbar is used in this example to drive the rotation of the square.
```
```suggestion
To demonstrate the effect of `scroll-timeline-axis`, a `horizontal` (non-default) scrollbar is used in this example to drive the rotation of the square.
```

estelle replied:
had to revert my earlier suggestion b/c the example still uses horizontal. we can change later.

estelle committed:
Update files/en-us/web/css/scroll-timeline-axis/index.md

estelle committed:
Update files/en-us/web/css/scroll-timeline-axis/index.md

estelle commented on files/en-us/web/css/scroll-timeline-axis/index.md:
```diff
@@ -68,12 +71,12 @@ The HTML for the example is shown below.
+The scroll axis is set by `scroll-timeline-axis: horizontal;` so that the _horizontal scrollbar_ position will determine the animation timeline.
 
-The height and width of the container are set to 200px and we also set the container to create scrollbars if it overflows (below we will use CSS on the "stretcher" element to ensure that it does overflow).
+The height and width of the container are set to `200px`. The container is also set to create scrollbars if it overflows (the CSS `height` and `width` rules on the `stretcher` element below do make the content overflow its container).
```
```suggestion
The height and width of the container are set to `200px`. The container is also set to create scrollbars if it overflows (the CSS `height` and `width` rules on the `stretcher` element below make the content overflow its container).
```

estelle committed:
Update files/en-us/web/css/scroll-timeline-axis/index.md

estelle commented on files/en-us/web/css/scroll-timeline/index.md:
```diff
@@ -26,55 +25,33 @@ This property is a shorthand for the following CSS properties:
 
+The `scroll-timeline` shorthand property can be applied to a container element as a combination of the `<scroll-timeline-name>` and `<scroll-timeline-axis>` values. At least one of the values must be specified. If both the values are specified, the order followed must be the `<scroll-timeline-name>` value followed by the `<scroll-timeline-axis>` value.
+
+> **Note:** When only one value is specified, the `scroll-timeline` shorthand property treats it as the `<scroll-timeline-name>` value. As a best practice, avoid using `<scroll-timeline-axis>` keywords as a `<scroll-timeline-name>` value.
```
```suggestion
> **Note:** When only one value is specified, the `scroll-timeline` shorthand property treats it as the `<scroll-timeline-name>` value. As a best practice, avoid using `<scroll-timeline-axis>` keyword values as a `<scroll-timeline-name>` value.
```

estelle committed:
Update files/en-us/web/css/scroll-timeline/index.md

captainbrosset commented:
> Hello @captainbrosset, I have a BCD-related question for this CSS `scroll-timeline` property in Chrome.
> 
> Per https://chromestatus.com/feature/6752840701706240, I see the implementation status is "Proposed". However, the feature is working on 111 and 112 beta with "Experimental Web Platform Features" enabled.

Not surprised, I've seen many cases of chromestatus not getting updated after a feature was first added to it. It's a manual update and there are cases where people simply forget to go back to it and make the necessary changes.

> ~[mdn/browser-compat-data#19086](https://github.com/mdn/browser-compat-data/issues/19086) mentions that the property was added to Chrome in version 40 (experimental).~
> 
> Would it be possible for you to check the version it was first added (even if in experimental mode) in Chrome and Edge? And if a different flag was needed in the older releases?

The easiest, and most correct, way to check this would be to ask the people who I know are the most knowledgeable about this right now on the Chrome team. @bramus and @argyleink, can you help answer Dipika's question above?

bramus commented:
> > ~[mdn/browser-compat-data#19086](https://github.com/mdn/browser-compat-data/issues/19086) mentions that the property was added to Chrome in version 40 (experimental).~
> > Would it be possible for you to check the version it was first added (even if in experimental mode) in Chrome and Edge? And if a different flag was needed in the older releases?
> 
> The easiest, and most correct, way to check this would be to ask the people who I know are the most knowledgeable about this right now on the Chrome team. @bramus and @argyleink, can you help answer Dipika's question above?

The earliest ScrollTimeline implementation that landed was support for [ScrollTimeline in Animation Worklets](https://bugs.chromium.org/p/chromium/issues/detail?id=719517). Using [my crbug-release-indicator extension](https://chrome.google.com/webstore/detail/crbug-release-indicator/ppbnemgliamfnlhhnclbibcpbnfdclnn) I see the implementation landed in version `61.0.3149.0`.

As for [the integration with Web Animations](https://bugs.chromium.org/p/chromium/issues/detail?id=916117), the first commits got included in Chromium `77.0.3830.0`.

As for [the CSS syntax](https://bugs.chromium.org/p/chromium/issues/detail?id=1074052), the first commits landed in `85.0.4165.0` but things only started to really work in version `86.0.4199.0` and got [enabled for experimental in `88.0.4307.0](https://bugs.chromium.org/p/chromium/issues/detail?id=1074052#c35)`

However, one big note with all this: this was for an older version of the spec that used the `@scroll-timeline` at-rule. Experimental support for the rewritten spec – that uses the CSS `scroll-timeline` and `view-timeline` properties – [first landed in Chrome 107](https://bugs.chromium.org/p/chromium/issues/detail?id=1344151#c4).

dipikabh commented:
Thanks a lot @captainbrosset for connecting me with the right folks.

Thank you @bramus for all the pointers. The release indicator extension is pretty useful 🙂.

> As for [the CSS syntax](https://bugs.chromium.org/p/chromium/issues/detail?id=1074052), the first commits landed in 85.0.4165.0 but things only started to really work in version 86.0.4199.0 and got [enabled for experimental in 88.0.4307.0](https://bugs.chromium.org/p/chromium/issues/detail?id=1074052#c35)

I believe Firefox also landed this property behind a preference around October last year (and then changed the preference name recently to match the updated spec). This update is getting indicated in the browser compat table via the following change: https://github.com/mdn/browser-compat-data/pull/19148/files.

To add a similar update for Chrome (and Edge) to the browser compat table, would it be accurate to specify 88 (experimental) as the original release version (and maybe have an additional entry for 107 to indicate update in syntax/spec)? However, it does not seem that the two longhand properties [`scroll-timeline-name`](https://developer.mozilla.org/en-US/docs/Web/CSS/scroll-timeline-name#browser_compatibility) and [`scroll-timeline-axis`](https://developer.mozilla.org/en-US/docs/Web/CSS/scroll-timeline-axis#browser_compatibility) were introduced in 88. So probably 107 (experimental) might be a better bet for any web developer looking for the three properties in Chrome?
[This 107 bug](https://bugs.chromium.org/p/chromium/issues/detail?id=1317765) also refers to the related properties `scroll-timeline-name` and `scroll-timeline-axis`.

The last of your links is for the [`view_timeline`](https://bugs.chromium.org/p/chromium/issues/detail?id=1344151#c4) property, which we haven't documented yet because of the WIP and experimental nature.

But it did help me find the [chrome bug](https://bugs.chromium.org/p/chromium/issues/detail?id=1357795) exactly matching the Firefox one for which I am updating the docs. From this bug:

> [scroll-timeline] Reverse scroll-timeline-axis/name
> 
> The grammar has changed to place the name first (to match
> view-timeline).
> 
> Note also that name/axis can no longer appear in any order, instead
> the name always comes first.

bramus commented:
> would it be accurate to specify 88 (experimental) as the original release version (and maybe have an additional entry for 107 to indicate update in syntax/spec)

SGTM
